src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -648,12 +669,76 @@ public void RemoveAnalyzerReference(string fullPath)
 
         #endregion
 
-        private void FileChangeContext_FileChanged(object sender, string fullFilePath)
+        private void DocumentFileChangeContext_FileChanged(object sender, string fullFilePath)
         {
             _sourceFiles.ProcessFileChange(fullFilePath);
             _additionalFiles.ProcessFileChange(fullFilePath);
         }
 
+        #region Metadata Reference Refreshing
+
+        private void FileReferenceChangeContext_FileChanged(object sender, string fullFilePath)
+        {
+            lock (_gate)
+            {
+                CancelOutstandingMetadataReferenceRefreshForFile_NoLock(fullFilePath);
+
+                var cancellationTokenSource = new CancellationTokenSource();
+                _metadataReferenceRefreshCancellationTokenSources.Add(fullFilePath, cancellationTokenSource);
+
+                Task.Delay(TimeSpan.FromSeconds(5), cancellationTokenSource.Token).ContinueWith(_ =>
+                {
+                    lock (_gate)
+                    {
+                        // We need to re-check the cancellation token source under the lock, since it might have been cancelled and restarted
+                        // due to another event
+                        cancellationTokenSource.Token.ThrowIfCancellationRequested();
+                        RefreshMetadataReferencesForFile_NoLock(fullFilePath);
+
+                        _metadataReferenceRefreshCancellationTokenSources.Remove(fullFilePath);
+                    }
+                }, cancellationTokenSource.Token, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+            }
+        }
+
+        private void RefreshMetadataReferencesForFile_NoLock(string fullFilePath)
+        {
+            // Since all adds/removals of references for this project happen under our lock, it's safe to do this
+            // check without taking the main workspace lock.
+            var project = _workspace.CurrentSolution.GetProject(Id);
+
+            foreach (var portableExecutableReference in project.MetadataReferences.OfType<PortableExecutableReference>())
+            {
+                if (portableExecutableReference.FilePath == fullFilePath)
```

Comment:
are we expecthing there are multiple metadata refernce with same file path? also, is it possible that there is no match?

Reply:
You can _theoretically_ have multiple references if you have multiple aliases and project system doesn't unify them into a single alias.

No files is also possible: if the file is notified on disk, and Windows has notified the VS file watcher, _but_ the file watcher hasn't notified us, _and_ simultaneously the assembly is removed, we might end up queueing a refresh when there isn't a reference anymore. We could fix this by adding as second walk across the list when we receive the file event, but that'd just add extra CPU to catch a scenario that's almost impossible to happen, and doing nothing further works fine. :smile:

Reply:
(I'll add a comment to this effect.)